File: Binary Search/k_Closest_numbers.cpp
class Solution {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        int n=arr.size();
        list<int> l;
        for(int i=0;i<min(k,n);i++){
            l.push_back(arr[i]);
        }
        
        for(int i=k;i<n;i++){
            if(abs(x-arr[i])<abs(x-l.front())){
                l.pop_front();
                l.push_back(arr[i]);
            }
        
        }
        vector<int> res(l.begin(),l.end());
        return res;
        
    }
};

File: Binary Search/k_Closest_numbers.py
class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        if x <= arr[0]:
            return arr[:k]
        if x >= arr[-1]:
            return arr[-k:]
        
        i = bisect.bisect_left(arr, x)
        low = max(0, i - k - 1)
        hi = min(len(arr) - 1, i + k - 1)
        
        while hi - low > k - 1:
            if x - arr[low] <= arr[hi] - x:
                hi -= 1
            else:
                low += 1
        return arr[low:hi+1]